src/dune_engine/build_system.ml

```diff
@@ -947,6 +953,35 @@ end = struct
       Path.Build.Set.to_list source_files_to_ignore
       |> Path.Source.Set.of_list_map ~f:Path.Build.drop_build_context_exn
     in
+
+    (* If a [.merlin] file is present in the [Promoted_to_delete] set but not in
+       the [Source_files_to_ignore] that means the rule that ordered its
+       promotion is no more valid. This would happen when upgrading to Dune 2.8
+       from ealier version without and building uncleaned projects. We delete
+       these leftover files here. *)
+    let merlin_file = ".merlin" in
+    let source_dir = Path.Build.drop_build_context_exn dir in
+    let merlin_in_src = Path.Source.(relative source_dir merlin_file) in
+    let source_files_to_ignore =
+      if
+        Path.Set.mem (Promoted_to_delete.load ()) (Path.source merlin_in_src)
+        && not (Path.Source.Set.mem source_files_to_ignore merlin_in_src)
+      then (
+        let path = Path.source merlin_in_src in
+        Log.info
```

Comment:
This seems pretty noisy. Do you think this message is important for the user to see?
(at the minimum I think the message should be hidden if the file is not on disk)

Reply:
It should happen only once when building an old project with Dune 2.8.3 and won't show unless --verbose is used. 

Once the file is not on disk it should also has been removed from the database and thus the message won't show.

Reply:
Ok, I didn't realize --verbose was necessary to show these.
As for being removed from the database, again that won't necessarily happen if you clean with git, for example. But I guess if you're cleaning with git you'll usually end up removing the _build directory as well, so it works out the same.